File: coach-academy/my-sheet-sol/sheet-5/ I - Ada and Queue.js
const fs = require("fs");
const input = fs
	.readFileSync(0, "utf8")
	.trim()
	.split("\n")
	.splice(1)
	.map((item) => item.split(" "));

const deque = [];
let reversed = false;
for (const arr of input) {
	let operation = arr[0];

	// instead of reversing the deque in O(n), we just swap the operations in O(1)
	if (reversed) {
		if (operation === "front") operation = "back";
		else if (operation === "back") operation = "front";
		else if (operation === "toFront") operation = "push_back";
		else if (operation === "push_back") operation = "toFront";
	}

	if ((operation === "front" || operation === "back") && deque.length === 0) {
		console.log("No job for Ada?");
		continue;
	}

	if (operation === "front") console.log(deque.shift());
	else if (operation === "back") console.log(deque.pop());
	else if (operation === "reverse") reversed = !reversed;
	else if (operation === "toFront") deque.unshift(arr[1]);
	else if (operation === "push_back") deque.push(arr[1]);
}

File: coach-academy/my-sheet-sol/sheet-5/ I - Ada and Queue.cpp
#include <iostream>
#include <queue>
using namespace std;

#define fast_io                \
  ios::sync_with_stdio(false); \
  cin.tie(nullptr);
#define endl '\n'

int main()
{
  fast_io;
  long long q;
  cin >> q;

  deque<short> dq;
  string op;
  short num;
  bool reversed = false;
  for (long long i = 0; i < q; i++)
  {
    cin >> op;

    if (reversed)
    {
      if (op == "toFront")
        op = "push_back";
      else if (op == "push_back")
        op = "toFront";
      else if (op == "front")
        op = "back";
      else if (op == "back")
        op = "front";
    }

    if (op == "toFront")
    {
      cin >> num;
      dq.push_front(num);
    }
    else if (op == "push_back")
    {
      cin >> num;
      dq.push_back(num);
    }
    else if (op == "front" && dq.size())
    {
      cout << dq.front() << endl;
      dq.pop_front();
    }
    else if (op == "back" && dq.size())
    {
      cout << dq.back() << endl;
      dq.pop_back();
    }
    else if (op == "reverse")
      reversed = !reversed;
    else
      cout << "No job for Ada?" << endl;
  }

  return 0;
}
